### DISCUSSION TITLE
Running faster using Bun instead of Node.js & some 💡 ideas

### DISCUSSION THREAD

**Discussion by africa552036012020255252355634346346346** [jeremyckahn/chitchatter#162] [General]
Been experimenting with [Bun](https://bun.sh/) ([quick video explainer](https://www.youtube.com/watch?v=dWqNgzZwVJQ)) a little bit and it seems to be running fine, even without changing anything in the code, just swapping out `npm` for `bun` (`bun start` etc.). Can also deploy on Cloudflare pages using Bun, [guide](https://gist.github.com/Hebilicious/88e5a444f42b8dc09fb86dfa865c6ed3).

I haven't timed it but it feels faster on my wooden pc and chat works just as usual (inspect console when running tho, some things may not load). It should also be possible to use Bun instead of Webpack, Im not too familiar with this field yet tho.

Few more random ideas: (written here so they're not forgotten, will be testing some when i get on my carbon fiber pc)
+ Switching to [Preact](https://preactjs.com/) instead of React, no clue if it works or if it's worth it.

+ If the viewport can't fit the menu, chat and peerlist at once (mobile devices) simply activate a `:focus` css tag that only allows either the menu, chat or peerlist to be focused at once, blurring the chat when either of the other is focused (+ something like `overflow:hidden` on the chat to prevent odd scrollbars / squished chat).

+ Peerlist (& the entire page) never extending beyond the initial viewport (fixing [this](https://github.com/jeremyckahn/chitchatter/issues/136)) (& tiny scroll in there Discord style for when having many connections if it's not there already).

+ Can [Deno KV](https://deno.com/kv) (beta open for everyone now) / [Cloudflare KV](https://www.cloudflare.com/developer-platform/workers-kv/) make for a simple active public rooms list (?) (& possibly even offline notifications ([ntfy](https://ntfy.sh/) integration needed (?)) + persistent (timed autodelete) messages, letting users bring their own KV keys (?)).

+ LLM integration where users can simply shove in their API keys (&/ something like [openrouter](https://openrouter.ai/docs)) 
plus parameters to define interactive bots, (overkill: [AgentGPT](https://github.com/reworkd/AgentGPT) &/ [MetaGPT](https://github.com/geekan/MetaGPT)-like integration; spawn and run entire software companies directly from the chat).

+ $ & Crypto integration, vision to have users be one or two clicks away from getting paid & sending payments in their chosen currencies (anonymously if desired) and for devs to be receiving a % cut of exchange fees. Even something like a simple [Trocador](https://trocador.app/en/) ([API](https://trocador.app/en/docs/)) integration can work.

**Comment by jeremyckahn** [comment]
Hi @africa552036012020255252355634346346346, thanks for sharing your ideas. I'll respond to each one.

> Been experimenting with [Bun](https://bun.sh/) ([quick video explainer](https://www.youtube.com/watch?v=dWqNgzZwVJQ)) a little bit and it seems to be running fine, even without changing anything in the code, just swapping out `npm` for `bun` (`bun start` etc.). Can also deploy on Cloudflare pages using Bun, [guide](https://gist.github.com/Hebilicious/88e5a444f42b8dc09fb86dfa865c6ed3).
> 
> I haven't timed it but it feels faster on my wooden pc and chat works just as usual (inspect console when running tho, some things may not load). It should also be possible to use Bun instead of Webpack, Im not too familiar with this field yet tho.

I'd prefer to stick with Node/Webpack for now, as they are the established standards in their respective spaces. Generally speaking, I prefer to optimize for stability and ecosystem when it comes to tooling. If Bun ever becomes the mainstream solution, we can revisit this.

> Switching to [Preact](https://preactjs.com/) instead of React, no clue if it works or if it's worth it.

I'd prefer not to do that, as my understanding is that Preact is not compatible with the entire React ecosystem. I'd prefer to maintain that compatibility for Chitchatter.

> Peerlist (& the entire page) never extending beyond the initial viewport (fixing [this](https://github.com/jeremyckahn/chitchatter/issues/136)) (& tiny scroll in there Discord style for when having many connections if it's not there already).

That seems like a great idea. I'd be open to accepting a PR that implements this.

> Can [Deno KV](https://deno.com/kv) (beta open for everyone now) / [Cloudflare KV](https://www.cloudflare.com/developer-platform/workers-kv/) make for a simple active public rooms list (?) (& possibly even offline notifications ([ntfy](https://ntfy.sh/) integration needed (?)) + persistent (timed autodelete) messages, letting users bring their own KV keys (?)).

I don't think this functionality is appropriate for Chitchatter. It's critical for privacy that room names remain unlisted, so I don't see a place for any kind of live room lists in this project.

> LLM integration where users can simply shove in their API keys (&/ something like [openrouter](https://openrouter.ai/docs))
plus parameters to define interactive bots, (overkill: [AgentGPT](https://github.com/reworkd/AgentGPT) &/ [MetaGPT](https://github.com/geekan/MetaGPT)-like integration; spawn and run entire software companies directly from the chat).

It's not clear to me what a practical use case for this might be. Can you explain what you're envisioning here?

> $ & Crypto integration, vision to have users be one or two clicks away from getting paid & sending payments in their chosen currencies (anonymously if desired) and for devs to be receiving a % cut of exchange fees. Even something like a simple [Trocador](https://trocador.app/en/) ([API](https://trocador.app/en/docs/)) integration can work.

I'd be open to accepting a PR that implements cryptocurrency transaction support within Chitchatter, provided that the functionality is unobtrusive, opt-in, and private.

**Reply by africa552036012020255252355634346346346** [reply]
I agree on all of your points. Stability, non-intrusiveness and optional opt-in is the way to go. I like the idea of having cool integrations be one or two clicks away as well as not being intrusive, but first prioritize important bug fixes of course.

To have active _public_ rooms be listed _publicly_ would guarantee to get quite some action going, while private rooms remain private as they should. Potential for abuse sure but this could be prevented with a simple captcha (in public rooms and only if & when needed) preventing bots + limiting to how fast one could send (public) messages. 
Also simply removing private room password requirement makes them into the current "public" rooms, while the truly private with password remain the same.

The LLM integration would be handy especially for having multiple people work together with multiple LLM's easily and possibly in a private peer to peer manner; self-hosted API's etc. (dangerous af tho; blackhat folks love the idea of easily accessible self-hosted uncensored LLM c o l l a b o r a t i o n).

An example:
"yo @ codegpt pls fix this github issue %link%" 
'Sure! %code%' 
"@ uncensoredgpt find exploits in that code"
'sending %this% GET request allows for remote code execution'
"@ codegpt ..."
'Apologies, here is the updated code %code%'
etc.

**Reply by jeremyckahn** [reply]
> To have active _public_ rooms be listed _publicly_ would guarantee to get quite some action going, while private rooms remain private as they should. Potential for abuse sure but this could be prevented with a simple captcha (in public rooms and only if & when needed) preventing bots + limiting to how fast one could send (public) messages.

That's a good point. Right now there are just Private and Public chat rooms. It could be worth expanding this model to support Private, Public, and Unlisted chat rooms. Public and Unlisted would work the same except that Public chat room names would be published to some central list where people could easily find them.

I'd be open to supporting this change if it was **extremely obvious** when a user is in a Public chat room versus an Unlisted chat room. Right now Chitchatter users (rightly) assume that Public chat rooms can't be found by strangers, so such a change would need to be introduced in a way that makes it clear that previously "Public" rooms are now discoverable.

> The LLM integration would be handy especially for having multiple people work together with multiple LLM's easily and possibly in a private peer to peer manner; self-hosted API's etc. (dangerous af tho; blackhat folks love the idea of easily accessible self-hosted uncensored LLM c o l l a b o r a t i o n).
>
> An example:
> "yo @ codegpt pls fix this github issue %link%"
> 'Sure! %code%'
> "@ uncensoredgpt find exploits in that code"
> 'sending %this% GET request allows for remote code execution'
> "@ codegpt ..."
> 'Apologies, here is the updated code %code%'
> etc.

Thanks for sharing the ideas you have in mind for this!

Ideally such an LLM would be easily self-hosted and neatly integrated right into Chitchatter. Something like https://webllm.mlc.ai/ seems like what we'd want, but the hardware requirements are quite extreme and not really accessible to most people. That exclusivity is not quite compatible with the ideals of Chitchatter. Given the exclusive nature of the current state of AI tech, it would worth exploring an **opt-in** integration with an external service like ChatGPT et al so long as the privacy implications are made clear to everyone affected. I would be open to supporting this change if anyone would like to lead the effort of implementing it.

**Comment by BDawg3327** [comment]
wsp
